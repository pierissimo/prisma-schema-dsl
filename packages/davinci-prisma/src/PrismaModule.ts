import { App, Module } from '@davinci/core'
import { ClassType } from '@davinci/reflector'
import fs from 'fs/promises'
import pathLib from 'path'
import { DataSourceProvider, Enum, Generator, Schema } from '@pmaltese/prisma-schema-generator'
import { exec } from 'child_process'
import { Logger } from 'pino'
import { PrismaGenerator } from './PrismaGenerator'

interface PrismaModuleOptions {
  models?: Array<ClassType>
  schemaGeneration?: {
    enabled: boolean
    path: string
  }
  clientGeneration?: {
    enabled: boolean
  }
  datasource: {
    name: string
    provider: DataSourceProvider
    url?: string
    environmentVariable?: string
  }
  generators?: Array<Generator>
  enums?: Array<Enum>
}

export class PrismaModule extends Module {
  app: App
  logger: Logger
  prismaSchema?: Schema
  prismaString?: string

  constructor(private options: PrismaModuleOptions) {
    super()
  }

  getModuleId() {
    return 'prisma'
  }

  async onRegister(app: App) {
    this.app = app
    this.logger = app.logger.child({ module: this.getModuleId() })

    if (this.options.schemaGeneration?.enabled) {
      await this.generateSchemaFile()
    }

    if (this.options.clientGeneration?.enabled) {
      await this.execPrismaGenerate()
    }
  }

  async generateSchemaFile() {
    const generator = new PrismaGenerator()

    const { schema, prismaString } = await generator
      .setDatasource(this.options.datasource)
      .setGenerators(this.options.generators ?? [])
      .setEnums(this.options.enums ?? [])
      .reflectModels(this.options.models ?? [])
      .generate()

    this.prismaString = prismaString
    this.prismaSchema = schema

    const relativePath = this.options.schemaGeneration?.path as string
    const path = pathLib.join(process.cwd(), relativePath)
    await fs.writeFile(path, prismaString)
    this.logger.info(`prisma definition file generated in: ${path}`)
  }

  execPrismaGenerate() {
    return new Promise((resolve, reject) => {
      exec('npx prisma generate', (error, stdout) => {
        if (error) {
          this.logger.error({ error })
          return reject(error)
        }

        this.logger.info(stdout)
        return resolve(stdout)
      })
    })
  }
}
