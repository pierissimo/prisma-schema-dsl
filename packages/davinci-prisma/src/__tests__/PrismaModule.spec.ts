import { DataSourceProvider } from '@pmaltese/prisma-schema-generator'
import { createApp } from '@davinci/core'
import fs from 'fs/promises'
import { PrismaModule } from '../PrismaModule'
import { prisma } from '../index'

describe('PrismaModule', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  test('should a', async () => {
    @prisma.model()
    class Customer {
      @prisma.field({ isId: true, required: true })
      id: string

      @prisma.field({ required: true })
      name: string
    }

    const prismaModule = new PrismaModule({
      datasource: {
        name: 'postgres',
        url: 'postgresql://',
        provider: DataSourceProvider.PostgreSQL,
      },
      models: [Customer],
      generators: [
        {
          name: 'client',
          provider: 'prisma-client-js',
          previewFeatures: ['fullTextSearch', 'fullTextIndex', 'clientExtensions'],
        },
      ],
      schemaGeneration: {
        enabled: true,
        path: '/path/to/prisma.schema',
      },
      clientGeneration: {
        enabled: true,
      },
    })

    const writeFileMock = jest.spyOn(fs, 'writeFile').mockImplementation()
    const execPrismaGenerateMock = jest
      .spyOn(prismaModule, 'execPrismaGenerate')
      .mockImplementation()

    const app = createApp()
    await app.registerModule(prismaModule)

    expect(writeFileMock).toHaveBeenCalled()
    const mockCallArguments = writeFileMock.mock.calls[0]
    expect(mockCallArguments[0]).toContain('/path/to/prisma.schema')
    expect(mockCallArguments[1]).toMatchSnapshot()

    expect(execPrismaGenerateMock).toHaveBeenCalled()
  })
})
