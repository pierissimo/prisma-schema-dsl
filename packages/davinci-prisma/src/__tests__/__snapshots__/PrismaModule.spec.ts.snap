// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrismaModule should a 1`] = `
"datasource postgres {
  provider = \\"postgresql\\"
  url      = \\"postgresql://\\"
}

generator client {
  provider        = \\"prisma-client-js\\"
  previewFeatures = [\\"fullTextSearch\\", \\"fullTextIndex\\", \\"clientExtensions\\"]
}

model Customer {
  id   String @id
  name String
}
"
`;
